dashboard/lib/pd/foorm/submission_analytics_parser.rb

```diff
@@ -0,0 +1,112 @@
+module Pd::Foorm
+  # Responsible for taking Submissions (ie, a response submitted by a user to one of our Forms)
+  # from our in-house survey system (Foorm)
+  # and processing them into a "one row per question answered" format
+  # for export to CSV, and, ultimately, to our analytics database (Redshift).
+  class SubmissionAnalyticsParser
+    include Constants
+
+    HEADERS = %w(
+      submission_id
+      question_name
+      matrix_item_name
+      response_value
+      response_text
+    )
+
+    # Loads 1000 records at a time to manage loading records into memory.
+    def self.reshape_all_submissions_and_export_to_csv
+      CSV.open('./test_submissions.csv', 'wb') do |csv|
+        csv << HEADERS
+
+        # Loads 1000 records at a time
+        ::Foorm::Submission.find_each do |submission|
+          reshaped_submission = reshape_submission(submission)
+
+          reshaped_submission.each do |answer|
+            row = answer.stringify_keys.values_at(*HEADERS)
+            csv << row
+          end
+        end
+      end
+    end
+
+    def self.reshape_submission(submission)
+      reshaped_submission_answers_with_metadata = []
+
+      submission_metadata = {submission_id: submission.id}
+
+      parsed_answers = JSON.parse(submission.answers)
+      reshaped_submission_answers = reshape_submission_answers(parsed_answers, submission.form)
+
+      reshaped_submission_answers.each do |reshaped_submission_answer|
+        reshaped_submission_answers_with_metadata << submission_metadata.merge(reshaped_submission_answer)
+      end
+
+      reshaped_submission_answers_with_metadata
+    end
+
+    # @param [Hash] parsed_answers a Hash of the answers attribute of a Foorm::Submission
+    # @param [Foorm::Form] form the form associated with the Foorm::Submission from which parsed_answers was derived
+    # @return [Array] array of hashes, with each hash representing an answer to export to CSV
+    def self.reshape_submission_answers(parsed_answers, form)
+      reshaped_submission_answers = []
+
+      parsed_answers.each do |question_name, answer|
+        reshaped_submission_answer = {question_name: question_name}
+        question_details = form&.get_question_details(question_name)
+
+        # If question isn't in the Form, return as-is.
+        # This is expected for metadata about the submission,
+        # which doesn't have an associated question in the Form.
```

Comment:
This seems a bit kludgy.  For my education, where is the extra metadata inserted?

Reply:
I believe on save of the Submission (ie, when the user completes a Form) to the DB. We include information like a variable containing the course about which the user completed the survey (eg, CS Principles vs. CS Discoveries), or other variables that influence what questions are shown / are substituted into question text.